### ISSUE TITLE
doOnListen converts to broadcast stream and is called on every listen, not just the first.

### ISSUE THREAD

**Issue by smkhalsa** [ReactiveX/rxdart#389]
The docs for `doOnListen` state that it 

> Invokes the given callback function when the stream is first listened to.

However, when I run the following...

```
final controller = StreamController();

final stream = controller.stream.doOnListen(() {
  print('listening');
});

stream.listen((data) {});
stream.listen((data) {});
```

The string gets printed twice.

> listening
> listening

This is unexpected for two reasons:
1. The stream has been converted to a broadcast stream, allowing multiple listens
2. The doOnListen callback gets called on each listen, not just on the first as described in the docs.

**Comment by smkhalsa** [comment]
@brianegan is this the expected behavior?

**Comment by hoc081098** [comment]
Haha. `stream` is single subscription stream

**Comment by frankpepermans** [comment]
Hmmm it shouldn't become a broadcast Stream :/

The fact that *every* listen triggers the handler, is as intended though, we need to improve the docs here.

**Comment by frankpepermans** [comment]
While writing some more tests for `doOnListen`, it seems that our implementation here is broken:

The following test hangs on the second listen:

```
test('doOnListen correctly allows subscribing multiple times on a broadcast stream', () {
      final controller = StreamController<int>.broadcast();
      final stream = controller.stream.doOnListen(() {
        // do nothing
      });

      stream.listen(null, onDone: () => print('done A'));
      stream.listen(null, onDone: () => print('done B')); // is being ignored

      expectLater(stream, emitsDone);
      expectLater(stream, emitsDone);

      controller.close();
    });
```
We keep a `subscriptions` `Map` internally to avoid listening to a `Stream` more than once for `doOnListen`, but this causes the second `onDone` method to be ignored.

**Comment by smkhalsa** [comment]
> The fact that every listen triggers the handler, is as intended though, we need to improve the docs here.

Ok, yes, the docs should be updated then.

If doOnListen cannot be used to invoke a callback only on the first listen, is there another operator that allows for this?

**Comment by frankpepermans** [comment]
No, any chance you can just capture that inside your handler, using a flag for example?

**Comment by frankpepermans** [comment]
should be fixed in 0.24.0-dev.1
